File: python/fastapi/tests/test_middleware.py
"""Tests for payskill-fastapi middleware.

Unit tests use FastAPI's TestClient (no running server needed) and a
local mock facilitator server for verify calls.
"""

from __future__ import annotations

import base64
import json
import threading
import time
from collections.abc import Generator
from typing import Any

import httpx
import pytest
import uvicorn
from fastapi import Depends, FastAPI, Request
from fastapi.testclient import TestClient

from payskill_fastapi import PaymentInfo, PayMiddleware, require_payment

# -- Mock facilitator ---------------------------------------------------------

_facilitator_behavior = {"mode": "valid"}


def _make_facilitator_app() -> FastAPI:
    app = FastAPI()

    @app.post("/verify")
    async def verify(request: Request) -> Any:
        mode = _facilitator_behavior["mode"]
        if mode == "valid":
            return {"isValid": True, "payer": "0x" + "cc" * 20}
        if mode == "invalid":
            return {"isValid": False, "invalidReason": "insufficient funds"}
        if mode == "error":
            from fastapi.responses import JSONResponse

            return JSONResponse({"error": "internal"}, status_code=500)
        return {"isValid": False}

    return app


class _FacilitatorServer:
    """Spin a facilitator on a background thread using uvicorn."""

    def __init__(self) -> None:
        self.app = _make_facilitator_app()
        self.server: uvicorn.Server | None = None
        self.thread: threading.Thread | None = None
        self.port = 0

    def start(self) -> None:
        import socket

        sock = socket.socket()
        sock.bind(("127.0.0.1", 0))
        self.port = sock.getsockname()[1]
        sock.close()

        config = uvicorn.Config(
            self.app,
            host="127.0.0.1",
            port=self.port,
            log_level="error",
        )
        self.server = uvicorn.Server(config)
        self.thread = threading.Thread(target=self.server.run, daemon=True)
        self.thread.start()
        # Wait for server to come up
        for _ in range(50):
            try:
                httpx.get(f"http://127.0.0.1:{self.port}/verify", timeout=0.1)
                return
            except httpx.HTTPError:
                time.sleep(0.05)
        raise RuntimeError("Facilitator server failed to start")

    def stop(self) -> None:
        if self.server:
            self.server.should_exit = True
        if self.thread:
            self.thread.join(timeout=2)

    @property
    def url(self) -> str:
        return f"http://127.0.0.1:{self.port}"


@pytest.fixture(scope="module")
def facilitator() -> Generator[_FacilitatorServer, None, None]:
    server = _FacilitatorServer()
    server.start()
    yield server
    server.stop()


# -- Helpers ------------------------------------------------------------------


def _make_payment_signature() -> str:
    payload = {
        "x402Version": 2,
        "accepted": {
            "scheme": "exact",
            "network": "eip155:8453",
            "amount": "10000",
        },
        "payload": {"signature": "0x" + "ab" * 65},
        "extensions": {},
    }
    return base64.b64encode(json.dumps(payload).encode()).decode()


# =============================================================================
# require_payment (provider) tests
# =============================================================================


class TestRequirePaymentNoHeader:
    def test_returns_402_without_header(self, facilitator: _FacilitatorServer) -> None:
        app = FastAPI()

        @app.get("/api/data")
        async def get_data(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=0.01,
                    settlement="tab",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url=facilitator.url,
                )
            ),
        ) -> Any:
            return {"data": "premium", "from": payment.from_address}

        client = TestClient(app)
        resp = client.get("/api/data")
        assert resp.status_code == 402

        pr_header = resp.headers.get("payment-required")
        assert pr_header, "PAYMENT-REQUIRED header must be present"

        decoded = json.loads(base64.b64decode(pr_header).decode())
        assert decoded["x402Version"] == 2
        assert len(decoded["accepts"]) == 1
        assert decoded["accepts"][0]["amount"] == "10000"
        assert decoded["accepts"][0]["extra"]["settlement"] == "tab"

        body = resp.json()
        assert body["detail"]["error"] == "payment_required"

    def test_amount_encoding_for_different_prices(self, facilitator: _FacilitatorServer) -> None:
        app = FastAPI()

        @app.get("/api/report")
        async def get_report(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=2.5,
                    settlement="direct",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url=facilitator.url,
                )
            ),
        ) -> Any:
            return {"report": "done"}

        client = TestClient(app)
        resp = client.get("/api/report")
        assert resp.status_code == 402

        pr_header = resp.headers.get("payment-required")
        decoded = json.loads(base64.b64decode(pr_header).decode())
        assert decoded["accepts"][0]["amount"] == "2500000"
        assert decoded["accepts"][0]["extra"]["settlement"] == "direct"


class TestRequirePaymentValid:
    def test_passes_through_with_payment_info(self, facilitator: _FacilitatorServer) -> None:
        _facilitator_behavior["mode"] = "valid"
        app = FastAPI()

        @app.get("/api/data")
        async def get_data(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=0.01,
                    settlement="tab",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url=facilitator.url,
                )
            ),
        ) -> Any:
            return {
                "paid_by": payment.from_address,
                "amount": payment.amount,
                "settlement": payment.settlement,
                "verified": payment.verified,
            }

        client = TestClient(app)
        resp = client.get(
            "/api/data",
            headers={"payment-signature": _make_payment_signature()},
        )
        assert resp.status_code == 200
        body = resp.json()
        assert body["paid_by"] == "0x" + "cc" * 20
        assert body["amount"] == 10000
        assert body["settlement"] == "tab"
        assert body["verified"] is True


class TestRequirePaymentInvalid:
    def test_returns_402_on_invalid_payment(self, facilitator: _FacilitatorServer) -> None:
        _facilitator_behavior["mode"] = "invalid"
        app = FastAPI()

        @app.get("/api/data")
        async def get_data(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=0.01,
                    settlement="tab",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url=facilitator.url,
                )
            ),
        ) -> Any:
            return {"data": "should not reach"}

        client = TestClient(app)
        resp = client.get(
            "/api/data",
            headers={"payment-signature": _make_payment_signature()},
        )
        assert resp.status_code == 402
        body = resp.json()
        assert "insufficient funds" in body["detail"]["message"]

    def test_returns_402_on_malformed_signature(self, facilitator: _FacilitatorServer) -> None:
        app = FastAPI()

        @app.get("/api/data")
        async def get_data(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=0.01,
                    settlement="tab",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url=facilitator.url,
                )
            ),
        ) -> Any:
            return {"data": "should not reach"}

        client = TestClient(app)
        resp = client.get(
            "/api/data",
            headers={"payment-signature": "not-valid-base64!!!"},
        )
        assert resp.status_code == 402
        body = resp.json()
        assert "decode failed" in body["detail"]["message"]


class TestRequirePaymentFacilitatorDown:
    def test_503_when_unreachable_closed(self) -> None:
        app = FastAPI()

        @app.get("/api/data")
        async def get_data(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=0.01,
                    settlement="tab",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url="http://127.0.0.1:19999",
                    fail_mode="closed",
                )
            ),
        ) -> Any:
            return {"data": "should not reach"}

        client = TestClient(app)
        resp = client.get(
            "/api/data",
            headers={"payment-signature": _make_payment_signature()},
        )
        assert resp.status_code == 503
        body = resp.json()
        assert body["detail"]["error"] == "facilitator_unavailable"

    def test_passthrough_when_unreachable_open(self) -> None:
        app = FastAPI()

        @app.get("/api/data")
        async def get_data(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=0.01,
                    settlement="tab",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url="http://127.0.0.1:19999",
                    fail_mode="open",
                )
            ),
        ) -> Any:
            return {
                "data": "passed through",
                "verified": payment.verified,
            }

        client = TestClient(app)
        resp = client.get(
            "/api/data",
            headers={"payment-signature": _make_payment_signature()},
        )
        assert resp.status_code == 200
        body = resp.json()
        assert body["data"] == "passed through"
        assert body["verified"] is True

    def test_402_when_facilitator_returns_500(self, facilitator: _FacilitatorServer) -> None:
        _facilitator_behavior["mode"] = "error"
        app = FastAPI()

        @app.get("/api/data")
        async def get_data(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=0.01,
                    settlement="tab",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url=facilitator.url,
                    fail_mode="closed",
                )
            ),
        ) -> Any:
            return {"data": "should not reach"}

        client = TestClient(app)
        resp = client.get(
            "/api/data",
            headers={"payment-signature": _make_payment_signature()},
        )
        # 500 from facilitator → isValid=False → 402
        assert resp.status_code == 402


class TestRequirePaymentFreeRoute:
    def test_free_route_unaffected(self, facilitator: _FacilitatorServer) -> None:
        app = FastAPI()

        @app.get("/api/health")
        async def health() -> Any:
            return {"ok": True}

        @app.get("/api/data")
        async def get_data(
            payment: PaymentInfo = Depends(
                require_payment(
                    price=0.01,
                    settlement="tab",
                    provider_address="0x" + "bb" * 20,
                    facilitator_url=facilitator.url,
                )
            ),
        ) -> Any:
            return {"data": "premium"}

        client = TestClient(app)
        # Free route - 200
        free_resp = client.get("/api/health")
        assert free_resp.status_code == 200

        # Paid route - 402
        paid_resp = client.get("/api/data")
        assert paid_resp.status_code == 402


# =============================================================================
# PayMiddleware (consumer) tests
# =============================================================================


class TestPayMiddleware:
    def test_attaches_pay_context(self) -> None:
        """request.state.pay has fetch and wallet."""
        from payskill import Wallet

        # Use a dummy wallet - we don't make real calls in this test
        test_key = "0x" + "ab" * 32
        wallet = Wallet(private_key=test_key, testnet=True)

        app = FastAPI()
        app.add_middleware(PayMiddleware, wallet=wallet)

        @app.get("/test")
        async def test_route(request: Request) -> Any:
            pay = request.state.pay
            return {
                "has_fetch": callable(pay.fetch),
                "has_wallet": pay.wallet is not None,
                "wallet_address": pay.wallet.address,
            }

        client = TestClient(app)
        resp = client.get("/test")
        assert resp.status_code == 200
        body = resp.json()
        assert body["has_fetch"] is True
        assert body["has_wallet"] is True
        assert body["wallet_address"].startswith("0x")

        wallet.close()

    def test_accepts_budget_options(self) -> None:
        """Middleware accepts max_per_request and max_total options."""
        from payskill import Wallet

        test_key = "0x" + "ab" * 32
        wallet = Wallet(private_key=test_key, testnet=True)

        app = FastAPI()
        app.add_middleware(
            PayMiddleware,
            wallet=wallet,
            max_per_request=1.00,
            max_total=50.00,
        )

        @app.get("/test")
        async def test_route(request: Request) -> Any:
            pay = request.state.pay
            return {"total_spent": pay.fetch.total_spent}

        client = TestClient(app)
        resp = client.get("/test")
        assert resp.status_code == 200
        assert resp.json()["total_spent"] == 0

        wallet.close()

File: typescript/express/tests/test_middleware.ts
/**
 * Tests for @pay-skill/express middleware.
 *
 * Unit tests use mocked fetch and in-process Express servers.
 * No external dependencies required.
 */

import { describe, it, before, after } from "node:test";
import assert from "node:assert/strict";
import http from "node:http";
import express from "express";
import { payMiddleware, requirePayment } from "../src/index.js";
import type { PaymentInfo } from "../src/types.js";

// ── Test helpers ─────────────────────────────────────────────────────

/** Start an Express app on a random port, return its URL and close fn. */
function serve(app: express.Express): Promise<{ url: string; close: () => void }> {
  return new Promise((resolve) => {
    const server = app.listen(0, () => {
      const addr = server.address();
      if (!addr || typeof addr === "string") throw new Error("bad address");
      const url = `http://127.0.0.1:${addr.port}`;
      resolve({ url, close: () => server.close() });
    });
  });
}

/**
 * Build a base64-encoded PAYMENT-REQUIRED header value.
 * Matches the x402 V2 format that requirePayment produces.
 */
function buildPaymentHeader(
  amount: string,
  payTo: string,
  settlement: string,
  facilitatorUrl: string,
): string {
  const requirements = {
    x402Version: 2,
    accepts: [
      {
        scheme: "exact",
        network: "eip155:8453",
        amount,
        asset: "0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913",
        payTo,
        maxTimeoutSeconds: 60,
        extra: { settlement, facilitator: facilitatorUrl },
      },
    ],
    resource: { url: "/api/data", mimeType: "application/json" },
    extensions: {},
  };
  return Buffer.from(JSON.stringify(requirements)).toString("base64");
}

/**
 * Build a fake PAYMENT-SIGNATURE header.
 * Real payments require wallet signing — this is just for provider test flow.
 */
function buildPaymentSignature(): string {
  const payload = {
    x402Version: 2,
    accepted: {
      scheme: "exact",
      network: "eip155:8453",
      amount: "10000",
    },
    payload: { signature: "0x" + "ab".repeat(65) },
    extensions: {},
  };
  return Buffer.from(JSON.stringify(payload)).toString("base64");
}

// ── Consumer middleware tests (payMiddleware) ─────────────────────────

describe("payMiddleware (consumer)", () => {
  // payMiddleware needs a real Wallet. Since we can't construct one in
  // unit tests without a key, we test the attachment behavior with a
  // mock wallet-like object.
  it("attaches req.pay with fetch and wallet", async () => {
    const app = express();

    // Create a minimal mock that satisfies the type
    const mockWallet = { address: "0x" + "aa".repeat(20) } as any;
    const mockFetch = (() => {}) as any;

    // Use a thin middleware that mimics payMiddleware behavior
    app.use((req, _res, next) => {
      req.pay = { fetch: mockFetch, wallet: mockWallet };
      next();
    });

    app.get("/test", (req, res) => {
      res.json({
        hasPay: !!req.pay,
        hasFetch: typeof req.pay?.fetch === "function",
        hasWallet: !!req.pay?.wallet,
        walletAddress: req.pay?.wallet?.address,
      });
    });

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/test`);
      const body = await resp.json() as any;
      assert.equal(body.hasPay, true);
      assert.equal(body.hasFetch, true);
      assert.equal(body.hasWallet, true);
      assert.equal(body.walletAddress, "0x" + "aa".repeat(20));
    } finally {
      close();
    }
  });
});

// ── Provider middleware tests (requirePayment) ───────────────────────

describe("requirePayment (provider)", () => {
  // Mock facilitator server
  let facilitatorUrl: string;
  let facilitatorClose: () => void;
  let facilitatorBehavior: "valid" | "invalid" | "error" | "timeout";

  before(async () => {
    const facilitator = express();
    facilitator.use(express.json());

    facilitator.post("/verify", (req, res) => {
      switch (facilitatorBehavior) {
        case "valid":
          res.json({ isValid: true, payer: "0x" + "cc".repeat(20) });
          break;
        case "invalid":
          res.json({ isValid: false, invalidReason: "insufficient funds" });
          break;
        case "error":
          res.status(500).json({ error: "internal" });
          break;
        case "timeout":
          // Don't respond — let it hang
          break;
      }
    });

    const server = await serve(facilitator);
    facilitatorUrl = server.url;
    facilitatorClose = server.close;
    facilitatorBehavior = "valid"; // default
  });

  after(() => {
    facilitatorClose();
  });

  it("returns 402 when no payment header is present", async () => {
    const app = express();
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl,
      }),
      (_req, res) => res.json({ data: "premium" }),
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/data`);
      assert.equal(resp.status, 402);

      // Check PAYMENT-REQUIRED header exists
      const prHeader = resp.headers.get("payment-required");
      assert.ok(prHeader, "PAYMENT-REQUIRED header must be present");

      // Decode and validate structure
      const decoded = JSON.parse(
        Buffer.from(prHeader, "base64").toString("utf-8"),
      );
      assert.equal(decoded.x402Version, 2);
      assert.equal(decoded.accepts.length, 1);
      assert.equal(decoded.accepts[0].amount, "10000"); // $0.01 = 10000 micro
      assert.equal(decoded.accepts[0].extra.settlement, "tab");

      // Check JSON body
      const body = await resp.json() as any;
      assert.equal(body.error, "payment_required");
    } finally {
      close();
    }
  });

  it("returns 402 with correct amount for different prices", async () => {
    const app = express();
    app.get(
      "/api/report",
      requirePayment({
        price: 2.5,
        settlement: "direct",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl,
      }),
      (_req, res) => res.json({ report: "done" }),
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/report`);
      assert.equal(resp.status, 402);

      const prHeader = resp.headers.get("payment-required")!;
      const decoded = JSON.parse(
        Buffer.from(prHeader, "base64").toString("utf-8"),
      );
      assert.equal(decoded.accepts[0].amount, "2500000"); // $2.50
      assert.equal(decoded.accepts[0].extra.settlement, "direct");
    } finally {
      close();
    }
  });

  it("passes through with req.payment on valid payment", async () => {
    facilitatorBehavior = "valid";
    const app = express();
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl,
      }),
      (req, res) => {
        const payment = req.payment as PaymentInfo;
        res.json({
          data: "premium",
          paidBy: payment.from,
          amount: payment.amount,
          settlement: payment.settlement,
          verified: payment.verified,
        });
      },
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/data`, {
        headers: { "PAYMENT-SIGNATURE": buildPaymentSignature() },
      });
      assert.equal(resp.status, 200);

      const body = await resp.json() as any;
      assert.equal(body.data, "premium");
      assert.equal(body.paidBy, "0x" + "cc".repeat(20)); // from mock facilitator
      assert.equal(body.amount, 10000);
      assert.equal(body.settlement, "tab");
      assert.equal(body.verified, true);
    } finally {
      close();
    }
  });

  it("returns 402 on invalid payment", async () => {
    facilitatorBehavior = "invalid";
    const app = express();
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl,
      }),
      (_req, res) => res.json({ data: "should not reach" }),
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/data`, {
        headers: { "PAYMENT-SIGNATURE": buildPaymentSignature() },
      });
      assert.equal(resp.status, 402);

      const body = await resp.json() as any;
      assert.equal(body.error, "payment_required");
      assert.ok(body.message.includes("insufficient funds"));
    } finally {
      close();
    }
  });

  it("returns 402 on malformed PAYMENT-SIGNATURE", async () => {
    const app = express();
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl,
      }),
      (_req, res) => res.json({ data: "should not reach" }),
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/data`, {
        headers: { "PAYMENT-SIGNATURE": "not-valid-base64!!!" },
      });
      assert.equal(resp.status, 402);

      const body = await resp.json() as any;
      assert.ok(body.message.includes("decode failed"));
    } finally {
      close();
    }
  });

  it("returns 503 when facilitator is unreachable (failMode=closed)", async () => {
    const app = express();
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        // Point to a port that doesn't exist
        facilitatorUrl: "http://127.0.0.1:19999",
        failMode: "closed",
      }),
      (_req, res) => res.json({ data: "should not reach" }),
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/data`, {
        headers: { "PAYMENT-SIGNATURE": buildPaymentSignature() },
      });
      assert.equal(resp.status, 503);

      const body = await resp.json() as any;
      assert.equal(body.error, "facilitator_unavailable");
    } finally {
      close();
    }
  });

  it("passes through when facilitator is unreachable (failMode=open)", async () => {
    const app = express();
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl: "http://127.0.0.1:19999",
        failMode: "open",
      }),
      (_req, res) => res.json({ data: "passed through" }),
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/data`, {
        headers: { "PAYMENT-SIGNATURE": buildPaymentSignature() },
      });
      assert.equal(resp.status, 200);

      const body = await resp.json() as any;
      assert.equal(body.data, "passed through");
    } finally {
      close();
    }
  });

  it("sets X-Pay-* headers on verified request", async () => {
    facilitatorBehavior = "valid";
    const app = express();
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl,
      }),
      (req, res) => {
        res.json({
          verified: req.headers["x-pay-verified"],
          from: req.headers["x-pay-from"],
          amount: req.headers["x-pay-amount"],
          settlement: req.headers["x-pay-settlement"],
          // payment-signature should be stripped
          hasPaymentSig: !!req.headers["payment-signature"],
        });
      },
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/data`, {
        headers: { "PAYMENT-SIGNATURE": buildPaymentSignature() },
      });
      assert.equal(resp.status, 200);

      const body = await resp.json() as any;
      assert.equal(body.verified, "true");
      assert.equal(body.from, "0x" + "cc".repeat(20));
      assert.equal(body.amount, "10000");
      assert.equal(body.settlement, "tab");
      assert.equal(body.hasPaymentSig, false);
    } finally {
      close();
    }
  });

  it("free routes are unaffected", async () => {
    const app = express();
    app.get("/api/health", (_req, res) => res.json({ ok: true }));
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl,
      }),
      (_req, res) => res.json({ data: "premium" }),
    );

    const { url, close } = await serve(app);
    try {
      // Free route — no 402
      const freeResp = await fetch(`${url}/api/health`);
      assert.equal(freeResp.status, 200);

      // Paid route — 402
      const paidResp = await fetch(`${url}/api/data`);
      assert.equal(paidResp.status, 402);
    } finally {
      close();
    }
  });

  it("returns 402 when facilitator returns 500 (failMode=closed)", async () => {
    facilitatorBehavior = "error";
    const app = express();
    app.get(
      "/api/data",
      requirePayment({
        price: 0.01,
        settlement: "tab",
        providerAddress: "0x" + "bb".repeat(20),
        facilitatorUrl,
        failMode: "closed",
      }),
      (_req, res) => res.json({ data: "should not reach" }),
    );

    const { url, close } = await serve(app);
    try {
      const resp = await fetch(`${url}/api/data`, {
        headers: { "PAYMENT-SIGNATURE": buildPaymentSignature() },
      });
      // Facilitator returned 500 → verify returns {isValid: false}
      assert.equal(resp.status, 402);
    } finally {
      close();
    }
  });
});
